### ISSUE TITLE
llvm formula is broken on CentOS 6.9

### ISSUE THREAD

**Issue by Char-Aznable** [Linuxbrew/homebrew-core#9611]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] have a problem with `brew install` (or `upgrade`, `reinstall`) a single formula? If it's a general `brew` problem please file this issue at Linuxbrew/brew: https://github.com/Linuxbrew/brew/issues/new/choose. If it's a tap (e.g. Brewsci/homebrew-bio) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ ] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:
- What you were trying to do (and why)
brew install llvm --with-toolchain --with-lldb --with-python@2

- What happened (include command output)
Multiple issues:
1. compiler-rt of llvm doesn't compile with the error:
```
In file included from /scratch/tmp/llvm-20180925-27468-svc990/llvm-7.0.0.src/projects/compiler-rt/lib/xray/xray_profile_collector.cc:15:
In file included from /scratch/tmp/llvm-20180925-27468-svc990/llvm-7.0.0.src/projects/compiler-rt/lib/xray/xray_profile_collector.h:21:
In file included from /scratch/tmp/llvm-20180925-27468-svc990/llvm-7.0.0.src/projects/compiler-rt/lib/xray/xray_function_call_trie.h:21:
In file included from /usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/memory:83:
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/shared_ptr.h:146:24: error: use of typeid requires -frtti
      { return __ti == typeid(_Deleter) ? &_M_del._M_del : 0; }
                       ^
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/shared_ptr.h:204:24: error: use of typeid requires -frtti
        return __ti == typeid(_Sp_make_shared_tag)
                       ^
[  5%] Built target RTSanitizerCommonCoverage.i386
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/shared_ptr.h:861:50: error: use of typeid requires -frtti
          void* __p = _M_refcount._M_get_deleter(typeid(__tag));
                                                 ^
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/shared_ptr.h:1005:52: error: use of typeid requires -frtti
    { return static_cast<_Del*>(__p._M_get_deleter(typeid(_Del))); }
```
I did a little bit of digging and found that the formula didn't take care of specifying the gcc toolchain used to build llvm. See [here](https://bobsteagall.com/2018/01/15/clang-builder/) and [here](https://www.allinonescript.com/index.php/questions/30352255/how-to-compile-llvm-against-a-custom-glibc) for details. 
2. After applying a fix (see below), I manage to get it build but the resulting clang is built in debug mode. I had to further add `-DCMAKE_C_FLAGS_RELEASE='-O3 -DNDEBUG' -DCMAKE_CXX_FLAGS_RELEASE='-O3 -DNDEBUG'` to bypass this issue. 
3. llvm's CMake system is picking up ancient version of libraries such as libcurses and backtrace from /usr/lib, which should not happen.

Overall, the fix that works for me is:
```
diff --git a/Formula/llvm.rb b/Formula/llvm.rb
index f28e527db6..cf639cfde1 100644
--- a/Formula/llvm.rb
+++ b/Formula/llvm.rb
@@ -253,6 +253,16 @@ class Llvm < Formula
     unless OS.mac?
       args << "-DLLVM_ENABLE_LIBCXXABI=ON"
       args << "-DCLANG_DEFAULT_CXX_STDLIB=libstdc++"
+      ENV["LDFLAGS"] = "-L#{HOMEBREW_PREFIX}/lib -Wl,--dynamic-linker=#{HOMEBREW_PREFIX}/lib/ld.so -Wl,-rpath,#{HOMEBREW_PREFIX}/lib" 
+      args << "-DGCC_INSTALL_PREFIX=#{Formula["gcc"].opt_prefix}"
+      args << "-DCMAKE_C_COMPILER=#{HOMEBREW_PREFIX}/bin/gcc"
+      args << "-DCMAKE_CXX_COMPILER=#{HOMEBREW_PREFIX}/bin/g++"
+      args << "-DCMAKE_C_FLAGS_RELEASE='-O3 -DNDEBUG'"
+      args << "-DCMAKE_CXX_FLAGS_RELEASE='-O3 -DNDEBUG'"
+      args << "-DCMAKE_LIBRARY_PATH=#{HOMEBREW_PREFIX}/lib"
+      args << "-DBacktrace_LIBRARY=#{HOMEBREW_PREFIX}/lib"
+      args << "-DBacktrace_LIBRARY=#{HOMEBREW_PREFIX}/lib"
+      args << "-DBacktrace_INCLUDE_DIRS=#{HOMEBREW_PREFIX}/include"
     end
```

- What you expected to happen
get a functional llvm

- Step-by-step reproduction instructions (by running `brew install` commands)
brew install llvm --with-toolchain --with-lldb --with-python@2

**Comment by xu-cheng** [comment]
https://github.com/Linuxbrew/homebrew-core/pull/9527#issuecomment-425994705

**Comment by iMichka** [comment]
I pushed some changes #9758, can you try again?

**Comment by Char-Aznable** [comment]
I see @xu-cheng submitted a new PR #9813 Which one should I try?

**Comment by xu-cheng** [comment]
I suggest you to try the llvm with the PR #9813, which restores a crucial patch to allow llvm finding brew glibc, include files and other libraries.

**Comment by iMichka** [comment]
All the patches have been merged. A new bottle has been built.

You can test again now.

**Comment by Char-Aznable** [comment]
New problem: build failed with (02.make):
```
/tmp/llvm-20181020-84880-6ddhou/llvm-7.0.0.src/projects/openmp/libomptarget/plugins/cuda/src/rtl.cpp:16:18: fatal error: cuda.h: No such file or directory
```
because (02.make.cc)
```
 g++-5 called with: -DTARGET_NAME=CUDA -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Domptarget_rtl_cuda_EXPORTS -I/tmp  /llvm-20181020-84880-6ddhou/llvm-7.0.0.src/build/projects/openmp/libomptarget/plugins/cuda -I/tmp/llvm-20181020-84880-6ddhou/llvm-7.0.0.src/projects/openmp/l  ibomptarget/plugins/cuda -I/home/aznb/.linuxbrew/include/libxml2 -I/tmp/llvm-20181020-84880-6ddhou/llvm-7.0.0.src/build/include -I/tmp/llv  m-20181020-84880-6ddhou/llvm-7.0.0.src/include -I/tmp/llvm-20181020-84880-6ddhou/llvm-7.0.0.src/projects/openmp/libomptarget/include -I/usr/local/cuda/includ  e -I/home/aznb/.linuxbrew/include/libelf -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++11 -Wall -Wextra -Wno-unused-paramete  r -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-maybe-uninitialized -Wdelete-non-virtual-dtor -Wno-comment -ffunc  tion-sections -fdata-sections -DNDEBUG -fPIC -o CMakeFiles/omptarget.rtl.cuda.dir/src/rtl.cpp.o -c /tmp/llvm-20181020-84880-6ddhou/llvm-7.0.0.src/projects/op  enmp/libomptarget/plugins/cuda/src/rtl.cpp
  superenv removed:  -I/usr/local/cuda/include -Werror=date-time -Wall -Wextra -Wwrite-strings -Wcast-qual -pedantic -Wdelete-non-virtual-dtor
  superenv added:    -pipe -w -O2 -march=native -isystem/home/aznb/.linuxbrew/include -I/home/aznb/.linuxbrew/opt/isl@0.1  8/include
```
I happen to be interested in the GPU offloading of OpenMP and libomptarget did pick up the cuda path in my system but superenv is over-supervising to remove it.

**Comment by Char-Aznable** [comment]
Aside from that, same error because of no gcc tool-chain specified.

**Comment by iMichka** [comment]
Regarding the cuda problem, did you modify the formula to explicitly activate the cuda features?
If yes, you may workaround the problem by using `--env=std` as a build option, so that superenv is disabled. We may need to find the cuda CMAKE flag and set it explicitly to OFF though.

Ideally we could also have a cuda formula, which would install the correct packages if needed.

**Comment by Char-Aznable** [comment]
CUDA is supposed to be picked up by default on machine that has it installed and it's the expected behavior to build LLVM's libomptarget. 
Setting `--env=std` gives whole chain of other problems, including the error as I posted at the beginning of this thread.

**Comment by Char-Aznable** [comment]
@xu-cheng Can you explain what #9813 is doing and how is it related to the error I post in the beginning? I don't think it's actually fixing the problem

**Comment by Char-Aznable** [comment]
GPU offloading is supposed to be set by default on machine with CUDA installed and it's the expected behavior of llvm's cmake and has nothing to do with linuxbrew formula. The only problem here is superenv is messing with things that it's not supposed to. 

@xu-cheng Can you explain what #9813 is doing and how it's related to the error I posted in the beginning? I don't think it has anything to do with it.

**Comment by Char-Aznable** [comment]
GPU offloading is supposed to be set by default on machine with CUDA installed and it's the expected behavior of llvm's cmake and has nothing to do with linuxbrew formula. The only problem here is superenv is messing with things that it's not supposed to. 

@xu-cheng Can you explain what #9813 is doing and how it's related to the error I posted in the beginning? I don't think it has anything to do with it.

**Comment by iMichka** [comment]
There is another thing that is missing, the glibc dependency for systems having older versions of glibc: #9993

@Char-Aznable it would also be interesting to know if `glibc@6` is working for you, to see if this is a llvm 7 specific bug.

**Comment by ndevenish** [comment]
I'm getting this and other errors today. I built gcc and glibc according to the Centos6 [instructions](https://github.com/Linuxbrew/brew/wiki/CentOS6). Pouring fails:

```==> Pouring llvm-7.0.0_2.x86_64_linux.bottle.1.tar.gz
tar: This does not look like a tar archive
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
Error: Failure while executing; `tar xf /home/mep23677/.cache/Homebrew/downloads/d097ed3446a3cdc9a931ab999008d2a60557ca194c5d7ae5f7e2b64c67500e72--llvm-7.0.0_2.x86_
64_linux.bottle.1.tar.gz -C /tmp/d20181025-24235-75ok1h` exited with 2. Here's the output:
tar: This does not look like a tar archive
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
```

And also I get the errors in compiler-rt specified in the first post.

It looks like the fix commits made in https://github.com/Linuxbrew/homebrew-core/pull/9813 and related have been history-rewritten at some point? My homebrew-core is up-to-date and I can't find the commits, and the `git log Formula/llvm.rb` doesn't show them:
```
$ git log --pretty=format:"%h%x09%an%x09%ad%x09%s" Formula/llvm.rb
673bf65 Michka Popoff   Thu Oct 25 11:35:52 2018 +0200  Merge branch homebrew/master into linuxbrew/master
d4c563c Michka Popoff   Wed Oct 24 15:23:23 2018 +0000  harfbuzz: update 2.0.0 bottle for Linuxbrew.
04f1955 Mike McQuaid    Sat Oct 13 13:14:02 2018 -0700  llvm: use :codesign requirement symbol.
ab427bc Viktor Szakats  Mon Oct 1 15:00:43 2018 +0000   llvm: secure url(s)
```

**Comment by Char-Aznable** [comment]
@iMichka What works for me is the `"-DGCC_INSTALL_PREFIX=#{Formula["gcc"].opt_prefix}"` cmake option. I provided documentation on why this needs to be set in order to compile llvm's compiler-rt

**Comment by xu-cheng** [comment]
> My system admin doesn't let me to publish gist logs anymore. 

This would make it hard for us to find the root problem. 

> Which part of the log do you want to see?

Could you paste the output of `brew config` and `brew doctor`? Can you check what compiler is used to build llvm (you can check brew log files with name similar to `0x.make.cc`)?

Judging from `/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/shared_ptr.h:146:24: error: use of typeid requires -frtti `, I think a wrong compiler is used. It seems that the old system compiler is used. In normal case, the brew gcc should be used. 

> What you were trying to do (and why)
brew install llvm --with-toolchain --with-lldb --with-python@2

You only need to run `brew install llvm --with-lldb`. The other two flags are intended for macOS.

**Comment by xu-cheng** [comment]
Another possible cause other than the wrong compiler I mentioned above is that the brew gcc in your system is not configured properly. So you could try: brew reinstall gcc.

**Comment by iMichka** [comment]
If a gist-log is not doable, a centos 6 Dockerfile with the steps to reproduce the problem would also be welcome.

**Comment by Char-Aznable** [comment]
@xu-cheng I think it would make sense for you to check your bottle build log to see if compiler-rt's xray was built using the OS's compiler than the linuxbrew one. Anyway, I was using the linuxbrew gcc 5.5.0_4. I don't think it's the problem with it because my proposed cmake options fixes the issue. Even in the original setting, llvm's other components were built using linuxbrew gcc but it's just that the llvm config decides to pick up the ancient OS gcc when switching to building compiler-rt. 


Brew config:
```
HOMEBREW_VERSION: 1.8.0
ORIGIN: https://github.com/Linuxbrew/brew
HEAD: 8f5de98bcc40afe402969dff6794739979b00a09
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Linuxbrew/homebrew-core
Core tap HEAD: 6b19fe741b8920d379641d41d126e4407b7f0f05
Core tap last commit: 24 hours ago
HOMEBREW_PREFIX: /home/aznb/.linuxbrew
HOMEBREW_REPOSITORY: /home/aznb/.linuxbrew
HOMEBREW_CELLAR: /home/aznb/.linuxbrew/Cellar
HOMEBREW_CACHE: /home/aznb/.cache
HOMEBREW_DEV_CMD_RUN: 1
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_LOGS: /home/aznb/.cache
HOMEBREW_NO_ANALYTICS: 1
CPU: 24-core 64-bit haswell
Homebrew Ruby: 2.3.7 => /home/aznb/.linuxbrew/Library/Homebrew/vendor/portable-ruby/2.3.7/bin/ruby
Clang: 7.0 build 700
Git: 2.16.2 => /home/aznb/.linuxbrew/bin/git
Curl: 7.19.7 => /usr/bin/curl
Kernel: Linux 2.6.32-696.18.7.el6.x86_64 x86_64 GNU/Linux
OS: CentOS release 6.9 (Final) (Final)
Host glibc: 2.12
/usr/bin/gcc: 4.4.7
glibc: 2.23
gcc: 5.5.0_4
xorg: 20170115_1
```
Running `brew doctor` takes forever on our crowded NFS system and I can't post it for the time being but last time I ran it was 2 weeks ago and there wasn't anything wrong with it. 

@iMichka  I can't do Docker either because it requires root to set up the daemon.

**Comment by xu-cheng** [comment]
> I think it would make sense for you to check your bottle build log to see if compiler-rt's xray was built using the OS's compiler than the linuxbrew one.

I don’t think this does any help to solve **your** problem. Further more, the current llvm formula works fine for me (also on CentOS 6), on CI and for other maintainers. In order to find the root cause, you really need to offer more information or demonstrate a way for us to reproduce the problem. For example, you can install docker in any other machine. 

> I don't think it's the problem with it because my proposed cmake options fixes the issue.

This fix will be unlikely to be accepted. Not only it makes the bottle unrelocable, it also completely bypasses brew’s superenv. Both of which are unwanted.

**Comment by Char-Aznable** [comment]
I have a ubuntu system that hits this problem but bypassing it because the OS gcc is of more modern version. The unintended phenomena is the switching of compiler when building different components of llvm. Do you see that in all your working cases? 

I don't think install docker in `any other machine` would help because it's not the machine that has the problem.


> > I think it would make sense for you to check your bottle build log to see if compiler-rt's xray was built using the OS's compiler than the linuxbrew one.
> 
> I don’t think this does any help to solve **your** problem. Further more, the current llvm formula works fine for me (also on CentOS 6), on CI and for other maintainers. In order to find the root cause, you really need to offer more information or demonstrate a way for us to reproduce the problem. For example, you can install docker in any other machine.

> 
> > I don't think it's the problem with it because my proposed cmake options fixes the issue.
> 
> This fix will be unlikely to be accepted. Not only it makes the bottle unrelocable, it also completely bypasses brew’s superenv. Both of which are unwanted.

Honestly, I don't care if you accept it or not. I was just posting for other people if they need it.

**Comment by Char-Aznable** [comment]
BTW, for anyone who's concerned, the bottle build not only misses lldb, but also GPU offloading of openmp and whole bunch of other features that I can't recall until I hit them again.

**Comment by ndevenish** [comment]
Replicated on docker, both the above failure to download the bottle and the failure to build compiler-rt. Logs from build at https://gist.github.com/ndevenish/936f39c5097e3746b9604457fa62d9d2, Dockerfile:
```dockerfile
FROM centos:6

RUN yum install -y gcc git automake make gcc-c++ pkgconfig patch

# Create user and target folder
RUN useradd -m brewser && \
    mkdir -p /opt/brew/rh6 && \
    chown brewser:brewser /opt/brew/rh6
USER brewser
WORKDIR /opt/brew/rh6

# Install linuxbrew base
RUN git clone https://github.com/Linuxbrew/brew.git .
ENV PATH="/opt/brew/rh6/bin:${PATH}"

# Install glibc per https://github.com/Linuxbrew/brew/wiki/CentOS6#install-linuxbrew-on-centos-6-without-sudo
RUN HOMEBREW_NO_AUTO_UPDATE=1 HOMEBREW_BUILD_FROM_SOURCE=1 brew install gcc --without-glibc
RUN HOMEBREW_NO_AUTO_UPDATE=1 brew install glibc
RUN HOMEBREW_NO_AUTO_UPDATE=1 brew remove gcc
RUN HOMEBREW_NO_AUTO_UPDATE=1 brew install gcc

# Install LLVM
RUN brew install llvm
```
Since I ran as a `docker build` I can't go in and grab the failure logs without rerunning (took >1.5 hours on my docker vm) but this looks exactly like the same error that I verified the logs against the original problem above.

**Comment by xu-cheng** [comment]
@ndevenish  Thanks for the information. 

```
==> Installing llvm
==> Downloading https://linuxbrew.bintray.com/bottles/llvm-7.0.0_2.x86_64_linux.bottle.1.tar.gz
==> Downloading from https://akamai.bintray.com/af/af0dd4ba9dfa9c219547790119d556969911c931f06ea2dc0f9ac0181e8bca72?__gda__=exp=1540555855~hmac=89e68c5c88557eecd28838d15e96b5c53ca13998872da519d65a0400a6192bef&response-content-disposition=attachment%3Bfilename%3D%22llvm-7.0.0_2.x86_64_linux.bottle.1.tar.gz%22&response-content-type=application%2Fgzip&requestInfo=U2FsdGVkX19jPrOmd8kuohd1nJjHv_EAByrIXCL12S-4MC76jQ7Niczo2AEMwzfFNEInzUSPwuqKgBoeqtpbLW9D8fD7pmNCXqFcUHqn8ubbA4JJP9KrtqQzBr-nqno-OAocfCHJ1E8mmJfB_SpZfA&response-X-Checksum-Sha1=129992ec1e442df73c73503f214652b9703836dd&response-X-Checksum-Sha2=af0dd4ba9dfa9c219547790119d556969911c931f06ea2dc0f9ac0181e8bca72
==> Pouring llvm-7.0.0_2.x86_64_linux.bottle.1.tar.gz
tar: This does not look like a tar archive
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
Error: Failure while executing; `tar xf /home/brewser/.cache/Homebrew/downloads/d097ed3446a3cdc9a931ab999008d2a60557ca194c5d7ae5f7e2b64c67500e72--llvm-7.0.0_2.x86_64_linux.bottle.1.tar.gz -C /tmp/d20181026-1-gri2iw` exited with 2. Here's the output:
tar: This does not look like a tar archive
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
```
@iMichka it seems that the current bottle is broken. We may need to rebuild the bottle.

**Comment by xu-cheng** [comment]
I was able to reproduce the problem using the instruction given by @ndevenish (Thanks!). 

The following is the gist log
https://gist.github.com/xu-cheng/aa78eaf7cc9779b60eb12432dc4a6ae8

I will try to figure out the problem when I find some free time next week.

**Comment by xu-cheng** [comment]
The relevant error message is 
```
cd /tmp/llvm-20181026-526-crk3dj/llvm-7.0.0.src/build/tools/clang/lib/ASTMatchers/Dynamic && /home/linuxbrew/.linuxbrew/Library/Homebrew/shims/linux/super/g++-5  -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/tmp/llvm-20181026-526-crk3dj/llvm-7.0.0.src/build/tools/clang/lib/ASTMatchers/Dynamic -I/tmp/llvm-20181026-526-crk3dj/llvm-7.0.0.src/tools/clang/lib/ASTMatchers/Dynamic -I/tmp/llvm-20181026-526-crk3dj/llvm-7.0.0.src/tools/clang/include -I/tmp/llvm-20181026-526-crk3dj/llvm-7.0.0.src/build/tools/clang/include -I/home/linuxbrew/.linuxbrew/include/libxml2 -I/tmp/llvm-20181026-526-crk3dj/llvm-7.0.0.src/build/include -I/tmp/llvm-20181026-526-crk3dj/llvm-7.0.0.src/include  -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-maybe-uninitialized -Wdelete-non-virtual-dtor -Wno-comment -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -fno-strict-aliasing -DNDEBUG     -o CMakeFiles/clangDynamicASTMatchers.dir/Registry.cpp.o -c /tmp/llvm-20181026-526-crk3dj/llvm-7.0.0.src/tools/clang/lib/ASTMatchers/Dynamic/Registry.cpp
{standard input}: Assembler messages:
{standard input}:288301: Error: unknown pseudo-op: `.st'
g++-5: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://github.com/Linuxbrew/homebrew-core/issues> for instructions.
```

It seems to be a problem of the gcc rather than llvm.

**Comment by iMichka** [comment]
> g++-5: internal compiler error: Killed (program cc1plus)

Not enough memory?

You could try to add this to the install block of the formula. Of course the number of cores may be adapted. But this is what we use on our CI:
> ENV["MAKEFLAGS"] = "-j2 -l2.0"

**Comment by ndevenish** [comment]
I'll re-run it with a higher memory limit and capturing the output logs tomorrow; on the RHEL6 real system it definitely had compile errors coming from the system GCC 4.4 header files.

**Comment by lunik1** [comment]
Hello, I think I am having the same issue on a RHEL 6 derived system. Here is the [gist](https://gist.github.com/d025bc3b33f147438647ae3f50b57d4a).

The error seeems to be a missing `<atomic>` header
```
/tmp/llvm-20181026-37140-cw34r/llvm-7.0.0.src/projects/compiler-rt/lib/xray/xray_x86_64.cc:15:10: fatal error: 'atomic' file not found
#include <atomic>
         ^~~~~~~~
1 error generated.
```
```
In file included from /tmp/llvm-20181026-37140-cw34r/llvm-7.0.0.src/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:16:
/tmp/llvm-20181026-37140-cw34r/llvm-7.0.0.src/projects/compiler-rt/lib/fuzzer/FuzzerInternal.h:23:10: fatal error: 'atomic' file not found
#include <atomic>
         ^~~~~~~~
```

**Comment by ndevenish** [comment]
I re-ran with a raised memory limit (8gb) and it failed with errors that aren't memory errors. Dockerfile is exactly the last line is `RUN brew install llvm || true` so that logs are preserved. I ran the gist-logs and it uploaded to https://gist.github.com/ndevenish/3ec024199d464bb40febc0002e3393f1, looks like the errors start in `02.make` line 2805:
```
[  1%] Building CXX object lib/xray/CMakeFiles/RTXray.x86_64.dir/xray_utils.cc.o
In file included from /tmp/llvm-20181030-97476-cupew0/llvm-7.0.0.src/projects/compiler-rt/lib/xray/xray_utils.cc:13:
In file included from /tmp/llvm-20181030-97476-cupew0/llvm-7.0.0.src/projects/compiler-rt/lib/xray/xray_utils.h:21:
In file included from /usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/utility:63:
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/stl_pair.h:255:27: error: no matching function for call to 'forward'
        (std::forward<_T1>(__x), std::forward<_T2>(__y));
                                 ^~~~~~~~~~~~~~~~~
```
So it definitely looks like it's picking up system headers.

**Comment by iMichka** [comment]
For those using bottles, I uploaded the tar file again, so the bottle should be back now.

**Comment by ndevenish** [comment]
Using my CentOS6 workflow fixes to binutils in #10088, the LLVM bottle now pours and appears to work, insofar as compiling basic test programs with clang seems fine.

**Comment by iMichka** [comment]
The `binutils` problem is probably unrelated, let's investigate this in #10088,

**Comment by ndevenish** [comment]
> The `binutils` problem is probably unrelated, let's investigate this in #10088,

Yes, sorry, didn't mean to imply so - just meant that "The bottle works" was subject to caveat of another fix. I've re-tested now that's fixed, and it the bottle works with the usual CentOS 6 instructions. Building is obviously still an issue.

**Comment by iMichka** [comment]
Current status:
- `binutils` problem fixed
- `llvm` bottle pouring fixed, the bottle is usable

Remaining issues:
- Wrong `gcc` is being picked up during build
- Out of memory errors (nothing we can do about)

**Comment by Char-Aznable** [comment]
I submitted PR#11074 for this based on https://llvm.org/docs/HowToCrossCompileBuiltinsOnArm.html#building-compiler-rt-builtins-for-arm

**Comment by iMichka** [comment]
Fixed by #11074
